bee-network/src/service/service.rs

```diff
@@ -0,0 +1,486 @@
+            let info = peerlist.info(&peer_id).unwrap();
+
+            let _ = senders.events.send(Event::PeerAdded { peer_id, info });
+
+            Ok(())
+        }
+        Err((peer_id, peer_info, mut e)) => {
+            // NB: This fixes an edge case where an in fact known peer connects before being added by the
+            // manual peer manager, and hence, as unknown. In such a case we simply update to the correct
+            // info (address, alias, relation).
+
+            if matches!(e, PeerError::PeerIsAdded(_))
+            // && peer_info.relation.is_known()
+            {
+                match peerlist.update_info(&peer_id, |info| *info = peer_info.clone()) {
+                    Ok(()) => {
+                        let _ = senders.events.send(Event::PeerAdded {
+                            peer_id,
+                            info: peer_info,
+                        });
+
+                        return Ok(());
+                    }
+                    Err(error) => e = error,
+                }
+            }
+
+            let _ = senders.events.send(Event::CommandFailed {
+                command: Command::AddPeer {
+                    peer_id,
+                    multiaddr: peer_info.address,
+                    // NOTE: the returned failed command now has the default alias, if none was specified originally.
+                    alias: Some(peer_info.alias),
+                    relation: peer_info.relation,
+                },
+                reason: e.clone(),
+            });
+
+            Err(e)
+        }
+    }
+}
+
+async fn remove_peer(peer_id: PeerId, senders: &Senders, peerlist: &PeerList) -> Result<(), PeerError> {
+    disconnect_peer(peer_id, senders, peerlist).await?;
+
+    let mut peerlist = peerlist.0.write().await;
+
+    match peerlist.remove(&peer_id) {
+        Ok(_peer_info) => {
+            let _ = senders.events.send(Event::PeerRemoved { peer_id });
+
+            Ok(())
+        }
+        Err(e) => {
+            let _ = senders.events.send(Event::CommandFailed {
+                command: Command::RemovePeer { peer_id },
+                reason: e.clone(),
+            });
+
+            Err(e)
+        }
+    }
+}
+
+async fn disconnect_peer(peer_id: PeerId, senders: &Senders, peerlist: &PeerList) -> Result<(), PeerError> {
+    let mut peerlist = peerlist.0.write().await;
+
+    // NB: We sent the `PeerDisconnected` event *before* we sent the shutdown signal to the stream writer task, so
+    // it can stop adding messages to the channel before we drop the receiver.
+
+    match peerlist.update_state(&peer_id, |state| state.set_disconnected()) {
+        Ok(Some(gossip_sender)) => {
+            let _ = senders.events.send(Event::PeerDisconnected { peer_id });
+
+            // Try to send the shutdown signal. It has to be a Vec<u8>, but it doesn't have to allocate.
+            let _ = gossip_sender.send(Vec::with_capacity(0));
```

Comment:
`Vec::with_capacity(0)` is just `Vec::new()`

Reply:
I know. I like how explicit it is in terms of self-documenting code. I'll keep it that way if you don't insisit.

Reply:
I don't see how this is useful? `Vec::new` is explicitly documented as being non-allocating. It's not conveying or documenting anything, it's just extra noise.